### ISSUE TITLE
Add config option to customize 'No such task' behavior

### ISSUE THREAD

**Issue by cldwalker** [babashka/babashka#1219]
Hi @borkdude 👋 

**Is your feature request related to a problem? Please describe.**
I'd like to be able to customize the behavior when a task is not found

**Describe the solution you'd like**
Support a top-level config option that takes the task not found and maybe any args e.g.

`:task-not-found (fn [task args] (run 'default-task))`

Some use cases I could use this for:

* Give a different error. Useful for commands like [lq](https://github.com/cldwalker/logseq-query/blob/main/bin/lq) that are basically wrappers around bb
* Allow users to auto-alias tasks on the fly if they'd like - [for example](https://github.com/cldwalker/bb-clis/blob/master/bin/bl)
* Fallback to a default command

**Describe alternatives you've considered**
I don't know of a way I can patch the internals since it seems hardcoded

**Additional context**

Happy to work on this if you agree it'd be helpful

**Comment by borkdude** [comment]
Yeah, we could have `:default` "hook"

```
:default (fn [{:keys [...]}] ...)
```

and we can add stuff to the input map. The default task then becomes responsible also for `bb not-found.clj` usage I could say?

**Comment by borkdude** [comment]
Btw, would it make sense to make a datascript pod, so that component becomes more re-usable for other bb users/scripts as well? There's already a datalevin pod btw:

https://github.com/babashka/pod-registry/blob/master/examples/datalevin.clj

**Comment by cldwalker** [comment]
> Maybe we should stick to the non-function approach:
> 
> ```
> :default (run 'default-task)
> ```
> 
> and we can add stuff to `(current-task)` which can be used in that default task. Args can already be accessed via `*command-line-args*`.

Sounds great 👍 Hopefully I'll get to it this weekend

**Comment by cldwalker** [comment]
> Btw, would it make sense to make a datascript pod, so that component becomes more re-usable for other bb users/scripts as well? There's already a datalevin pod btw:
> 
> https://github.com/babashka/pod-registry/blob/master/examples/datalevin.clj

That would definitely be nice. I didn't see much documentation on building releases and I don't have much personal time to do this. If I give this a try, what would be the easiest example to copy from?

**Comment by borkdude** [comment]
> If I give this a try, what would be the easiest example to copy from?

@cldwalker Probably buddy (recently updated) or datalevin (since that's close to datascript).

**Comment by ieugen** [comment]
I'm also looking for something like this. 
My use case is to have an easy way to call bb tasks from a central location. 
I would like to have a bb.edn where I can register tasks from other bb.edn in other directories. 

One way I can do that is to read bb.edn files and register tasks dynamically,  based on the tasks I find in those files. 

Simple way of doing this: 
```clojure
{:tasks {:requires ([clojure.string :as str])
         app {:doc "Without arguments, prints tasks in app - with arguments, invokes task in app dir"
              :task (if (empty? *command-line-args*)
                      (shell {:dir "app"} "bb tasks")
                      (shell {:dir "app"} (str "bb run " (str/join " " *command-line-args*))))}}}
```

Using a default function I could implement the logic based on the command line parameters. 
I could scan a list of directories (or all subdirectories) for bb.edn and register them as available tasks.

**Comment by borkdude** [comment]
Another alternative is to generate the `bb.edn` once based on the available directories, using a script.

**Comment by ieugen** [comment]
I'll think about it and maybe try it out. 
if the tasks are static yes, but then people would have to remember to re-generate the script when changes happen (maintenance).

**Comment by borkdude** [comment]
@ieugen That is true, but I expect the churn of subdirectories in a project to not be so high that this becomes annoying. Also there are ways to automate that by e.g. adding the bb.edn re-generation as a task dependency of other common ran tasks or as a git pre-commit hook, or both. I've done that e.g. here in clerk:

Installing the git commit hook, happens when invoking a task
https://github.com/nextjournal/clerk/blob/9a04cf38fd4d7c771071ddc14538da673cd32f4f/bb.edn#L75

And the actually pre-commit hook executes a task which refreshes a file (in this case a SHA, but in your case it would be the re-generated bb.edn):

https://github.com/nextjournal/clerk/blob/main/dev/pre-commit

**Comment by borkdude** [comment]
Another thought, what about:

```
bb app/task
```

which executes a task in directory app, from `app/bb.edn`?

Not entirely sure if that's a good idea or not, but currently it was already not possible to use slashes in task names, so it should at least not be breaking.

**Comment by borkdude** [comment]
And possibly `bb tasks` could also print the tasks in subdirs, but as separate groups.

```
bb tasks

foo Task in current dir

app/task2 Task in dir app
app/task1 Another task in dir app
```

**Comment by ieugen** [comment]
I think both approaches have merit.

**Comment by borkdude** [comment]
Let's continue the discussion about tasks in monorepos here: https://github.com/babashka/babashka/discussions/1044
It seems to be a slightly differerent use case as the `:default` fallback.

**Comment by just-sultanov** [comment]
It seems to me that there is no need to mix work on modules with the default task handler, as the user can create any logic within the default task. I have tried testing working with modules today, and it has been enough for me. The new feature with modules should be developed separately, regardless of the default task handler.

Sorry, I wanted to add a test and then submit a pull request, but I couldn't find any examples of how to properly prepare the `sci` context and other required things.

This example below shows how I added the default task handler:

```bash
diff --git a/src/babashka/impl/tasks.clj b/src/babashka/impl/tasks.clj
index 3fe45785..7b3e1831 100644
--- a/src/babashka/impl/tasks.clj
+++ b/src/babashka/impl/tasks.clj
@@ -282,7 +282,8 @@
         tasks (get bb-edn :tasks)
         enter (:enter tasks)
         leave (:leave tasks)
-        task (get tasks task-name)]
+        task (or (get tasks task-name)
+                 (:default tasks))]
     (binding [*print-meta* true]
       (if task
         (let [m? (map? task)
```

**Comment by borkdude** [comment]
I'm not sure what you are saying. If this summary correct?

If bb adds a fallback / default task option, people can do their own thing to solve the "modules" problem?

But if that is correct, then why do we even need this:

> The new feature with modules should be developed separately,

**Comment by just-sultanov** [comment]
I mean, if we want to add some internal functionality to work with modules, it's better to develop it as an independent feature. 

The default task handler will close some of the problems, e.g.:

If we add a default task handler, we can avoid adding unnecessary commands of the same type to `bb.edn` [like this](https://github.com/just-sultanov/bb-modules/blob/main/bb.edn#L61-L69) and our `bb.edn` can be simplified:

```clojure
{:tasks
 {:requires ([clojure.string :as str])

  :init (do
          (def root-path (or (System/getenv "PROJECT_ROOT_PATH") "."))

          (def modules
            {"module1" {:path (format "%s/modules/module1" root-path)
                        :description "Module 1"}
             "module2" {:path (format "%s/modules/module2" root-path)
                        :description "Module 2"}})

          (defn exec-task
            [module args]
            (let [dir (get-in modules [module :path])]
              (if (not (seq args))
                (shell {:dir dir} "bb tasks")
                (shell {:dir dir} (format "bb run %s" (str/join \space args)))))))

  :default {:task (let [task (current-task)
                        task-name (-> task :name name)]
                    (if (contains? modules task-name)
                      (exec-task task-name *command-line-args*)
                      (println (format "No such module: %s\n" task-name))))}}}

```
